spec/ordinary-and-exotic-object-behaviours.html

```diff
@@ -0,0 +1,326 @@
+        <emu-alg>
+          1. Let _keys_ be a new empty List.
+          1. Let _rec_ be _R_.[[RecordData]].
+          1. Assert: Type(_rec_) is Record.
+          1. Let _len_ be the length of _rec_.
+          1. For each pair _p_ of _rec_.[[Fields]], do
+            1. Let _key_ be _p_.[[Key]]
+            1. Assert: Type(_key_) is String
+            1. Add _key_ to _keys_
+          1. Return _keys_
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-recordcreate" aoid="RecordCreate">
+        <h1>RecordCreate ( _value_ )</h1>
+        <p>This abstract operation RecordCreate takes a _value_ argument. It is used to specify the creation of new Record exotic objects. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: Type(_value_) is a Record.
+            1. Assert: Type(_value_.[[Fields]]) is a List.
+            1. Let _R_ be ! MakeBasicObject(&laquo; [[RecordData]] &raquo;).
+            1. Set _R_'s essential internal methods to the definitions specified in <emu-xref href="#sec-record-exotic-objects"></emu-xref>.
+            1. Set _R_.[[Prototype]] to *null*.
+            1. Set _R_.[[RecordData]] to _value_.
+            1. OrdinaryPreventExtensions(_R_).
+            1. Return _R_
+            
+        </emu-alg>
+      </emu-clause>
+      
+      <emu-clause id="sec-recordhasproperty-r-p" aoid="RecordHasProperty">
+        <h1>RecordHasProperty ( _R_, _P_ )</h1>
+        <p>This abstract operation RecordHasProperty takes arguments _R_ and _P_. It performs the following steps when called:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is a Symbol, return *false*.
+          1. Let _rec_ be _R_.[[RecordData]].
+          1. Assert Type(_rec_) is a Record.
+          1. For each pair _p_ of _rec_.[[Fields]], do
+            1. Let _key_ be _p_.[[Key]]
+            1. If SameValue(_key_, _P_), return *true*.
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+      
+      <emu-clause id="sec-recordget-r-p" aoid="RecordGet">
+        <h1>RecordGet ( _R_, _P_ )</h1>
+        <p>This abstract operation RecordGet takes arguments _R_ and _P_. It performs the following steps when called:</p>
+        <emu-alg>
+          1. Assert: _R_ is an Object that has [[RecordData]] internal slot.
+          1. Assert: Type(_P_) is a String.
+          1. Let _rec_ be _R_.[[RecordData]].
+          1. Assert Type(_rec_) is a Record.
+          1. Let _len_ be the length of _rec_.
+          1. Let _resValue_ be *undefined*.
+          1. For each pair _p_ of _rec_.[[Fields]], do:
+            1. Let _key_ be _p_.[[Key]].
+            1. Let _val_ be _p_.[[Value]].
+            1. Assert: Type(_key_) is a String.
+            1. Assert: Type(_val_) is not an Object.
+            1. If SameValue(_key_, _P_), let _resValue_ be _val_.
```

Comment:
Note, it's fine to early-return from an algorithm, so you could "return _val_" here, and fall back to returning undefined below.